File: tests/klinger_volume_oscillator_test.py
import numpy as np
import wickra

from taflow import KlingerVolumeOscillator


def test_klinger_volume_oscillator_matches_wickra():
    rng = np.random.default_rng(29)
    close = 100.0 + np.cumsum(rng.normal(size=256))
    high = close + rng.uniform(0.2, 1.5, len(close))
    low = close - rng.uniform(0.2, 1.5, len(close))
    volume = rng.uniform(100.0, 1000.0, len(close))
    actual = KlingerVolumeOscillator().extend(high, low, close, volume).compute()
    expected = np.asarray(
        wickra.KVO().batch(high, low, close, volume), dtype=np.float64
    )

    # Wickra exposes the oscillator only. TAFlow's signal line is an additional
    # native lifecycle output and is covered by chunk/reset invariance below.
    np.testing.assert_allclose(actual[0], expected, equal_nan=True, atol=1e-10)


def test_klinger_volume_oscillator_chunked_reset():
    close = np.linspace(90.0, 130.0, 80)
    high = close + 1.0
    low = close - 1.0
    volume = np.full_like(close, 500.0)
    whole = KlingerVolumeOscillator(5, 8, 3).extend(high, low, close, volume)
    chunked = KlingerVolumeOscillator(5, 8, 3)
    chunked.extend(high[:31], low[:31], close[:31], volume[:31])
    chunked.extend(high[31:], low[31:], close[31:], volume[31:])
    for left, right in zip(whole.compute(), chunked.compute()):
        np.testing.assert_array_equal(left, right)
    assert chunked.reset() is chunked
    assert len(chunked) == 0

File: crates/taflow-core/src/indicators/klinger_volume_oscillator_test.rs
use super::klinger_volume_oscillator::KlingerVolumeOscillator;

#[test]
fn bulk_and_reset_replay_match_scalar_state() {
    let high: Vec<f64> = (0..128).map(|index| 100.0 + index as f64 + 1.0).collect();
    let low: Vec<f64> = high.iter().map(|value| value - 2.0).collect();
    let close: Vec<f64> = high.iter().map(|value| value - 1.0).collect();
    let volume: Vec<f64> = (0..128).map(|index| 1000.0 + index as f64).collect();
    let mut state = KlingerVolumeOscillator::new(5, 8, 3).unwrap();
    let mut oscillator = Vec::new();
    let mut signal = Vec::new();
    state
        .extend_slice_into(&high, &low, &close, &volume, &mut oscillator, &mut signal)
        .unwrap();
    let final_value = state.value();
    state.reset();
    let mut replay_oscillator = Vec::new();
    let mut replay_signal = Vec::new();
    state
        .extend_slice_into(
            &high,
            &low,
            &close,
            &volume,
            &mut replay_oscillator,
            &mut replay_signal,
        )
        .unwrap();
    assert_eq!(oscillator.len(), replay_oscillator.len());
    assert_eq!(signal.len(), replay_signal.len());
    for (left, right) in oscillator.iter().zip(replay_oscillator) {
        assert_eq!(left.to_bits(), right.to_bits());
    }
    for (left, right) in signal.iter().zip(replay_signal) {
        assert_eq!(left.to_bits(), right.to_bits());
    }
    assert_eq!(state.value(), final_value);
}
